packages/ifpack2/src/Ifpack2_Details_CrsArrays.hpp

```diff
@@ -0,0 +1,187 @@
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY SANDIA CORPORATION "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Michael A. Heroux (maherou@sandia.gov)
+//
+// ***********************************************************************
+//@HEADER
+*/
+
+#ifndef __IFPACK2_CRSARRAYS_DECL_HPP__ 
+#define __IFPACK2_CRSARRAYS_DECL_HPP__ 
+
+#include <Tpetra_RowMatrix.hpp>
+#include <Tpetra_CrsMatrix.hpp>
+#include <Kokkos_DefaultNode.hpp>
+#include <Kokkos_CrsMatrix.hpp>
+#include <Ifpack2_LocalFilter.hpp>
+#include <Ifpack2_ReorderFilter.hpp>
+
+namespace Ifpack2
+{
+namespace Details
+{
+
+//Utility for getting the local values, rowptrs and colinds (in Kokkos::Views) for any RowMatrix
+//Used by Fic, Filu and Fildl but may also be useful in other classes
+template<typename Scalar, typename LocalOrdinal, typename GlobalOrdinal, typename Node>
+struct CrsArrayReader
+{
+  typedef typename Node::device_type Device;
+  typedef typename Device::execution_space ExecSpace;
+  typedef Tpetra::RowMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node> TRowMatrix;
+  typedef Tpetra::CrsMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node> TCrsMatrix;
+  typedef Ifpack2::LocalFilter<TRowMatrix> Filter;
+  typedef Ifpack2::ReorderFilter<TRowMatrix> ReordFilter;
+  typedef Kokkos::CrsMatrix<Scalar, LocalOrdinal, ExecSpace> KCrsMatrix;
+  typedef Kokkos::View<LocalOrdinal*, ExecSpace> OrdinalArray;
+  typedef Kokkos::View<Scalar*, ExecSpace> ScalarArray;
+
+  //These functions are only meant to be used by Fic/Filu/Fildl for AdditiveSchwarz
+  //A's underlying type can either be a CrsFilter or a RowFilter
+  static void getValues(const TRowMatrix* A, ScalarArray& vals)
```

Comment:
Why not just have one function that returns all three arrays?  The current design requires that you iterate over the matrix's rows three times, once for each of the three arrays.